### PR TITLE
make testclient more precise

### PR DESCRIPTION
**Pull request by deads2k**
Updates the fake testclient to pass explicit verbs and resources to reaction functions.  It also makes the information more complete for update, list, and watch actions.

@mikedanese I remember you made changes in this area before, can you take a look?

### CODE REVIEW

File: pkg/client/testclient/actions.go [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+	return a.Verb
+}
+func (a ActionImpl) GetResource() string {
+	return a.Resource
+}
+func (a ActionImpl) GetSubresource() string {
+	return a.Subresource
+}
+func (a ActionImpl) Matches(verb, resource string) bool {
+	return verb == a.Verb && resource == a.Resource
```
Comment (mikedanese) [inline pkg/client/testclient/actions.go]: nit: make the verb check case insensitive?

File: pkg/client/testclient/fixture.go [hunk 2]
```diff
@@ -47,34 +47,43 @@ type ObjectRetriever interface {
...
 		if err != nil {
-			return nil, fmt.Errorf("unrecognized action %s: %v", resource, err)
+			return nil, fmt.Errorf("unrecognized action %s: %v", action.GetResource(), err)
 		}
+
 		// TODO: have mapper return a Kind for a subresource?
-		switch verb {
-		case "list", "search":
+		switch castAction := action.(type) {
+		// TODO for some reason fake_events.go uses a "search" verb.  I have no idea why.  Seems wrong
```
Comment (mikedanese) [inline pkg/client/testclient/fixture.go]: "search" doesn't appear in swagger docs. It should likely be a "get". :-1:

File: pkg/kubelet/kubelet_test.go [hunk 3]
```diff
@@ -2424,7 +2427,11 @@ func TestUpdateExistingNodeStatus(t *testing.T) {
 	if len(actions) != 2 {
 		t.Errorf("unexpected actions: %v", actions)
 	}
-	updatedNode, ok := actions[1].Value.(*api.Node)
+	updateAction, ok := actions[1].(testclient.UpdateAction)
+	if !ok {
+		t.Errorf("unexpected object type")
```
Comment (mikedanese) [inline pkg/kubelet/kubelet_test.go]: should be "unexpected action type" (and include the expected and actual action type)

File: pkg/kubelet/status_manager_test.go [hunk 4]
```diff
@@ -158,7 +160,7 @@ func TestSyncBatch(t *testing.T) {
 	if err != nil {
 		t.Errorf("unexpected syncing error: %v", err)
 	}
-	verifyActions(t, syncer.kubeClient, []string{"get-pod", "update-status-pod"})
+	verifyActions(t, syncer.kubeClient, [][3]string{{"get", "pods", ""}, {"update", "pods", "status"}})
```
Comment (mikedanese) [inline pkg/kubelet/status_manager_test.go]: why not pass in actions?

**Comment by deads2k** [comment]
Comments addressed.  Also added checks against casting nils from reaction funcs in a separate commit.

**Comment by mikedanese** [comment]
Thanks for the cleanup, this LGTM on green. Can you please squash?

**Comment by deads2k** [comment]
squashed.

**Comment by mikedanese** [comment]
@deads2k tests are failing.

**Comment by smarterclayton** [comment]
Changes look generally OK, but don't have time to give it a deeper look

File: pkg/client/testclient/fake_events.go [hunk 5]
```diff
@@ -30,48 +30,69 @@ type FakeEvents struct {
 	Fake *Fake
 }
 
-// Create makes a new event. Returns the copy of the event the server returns, or an error.
-func (c *FakeEvents) Create(event *api.Event) (*api.Event, error) {
-	obj, err := c.Fake.Invokes(FakeAction{Action: "get-event", Value: event.Name}, &api.Event{})
-	return obj.(*api.Event), err
+func nilEvent() *api.Event {
+	var ret *api.Event
+	return ret
```
Comment (kargakis) [inline pkg/client/testclient/fake_events.go]: why not explicitly return nil?

Comment (deads2k) [inline pkg/client/testclient/fake_events.go]: > why not explicitly return nil?

Because I lost my mind.  Fixed.

**Comment by deads2k** [comment]
Looks like go 1.4 is happy on shippable and go 1.3 is unhappy.  Any chance its a known problem?

**Comment by deads2k** [comment]
@mikedanese alright, shippable went green, but now Jenkins is red and I can't see the error.  Can you let me know what failed or re-kick if it looks like a flake?

**Comment by mikedanese** [comment]
looks like another flake. #12174. @k8s-bot ok to test.

**Comment by alex-mohr** [comment]
LGTM was before last commit, removing LGTM [contrib/submit-queue: auto remove LGTM]

**Comment by alex-mohr** [comment]
[Sorry, still a bug in submit-queue LGTM handling.  Re-applied.]

**Comment by alex-mohr** [comment]
@k8s-bot test this [contrib/submit-queue: candidate for merging]

**Pull request merged**